manualTests/AdapterExtensionsGetAllDevicesAsync/Program.cs

```diff
@@ -0,0 +1,56 @@
+﻿// See https://github.com/vestervang/DotNet-BlueZ/issues/8
+
+using vestervang.DotNetBlueZ;
+
+namespace AdapterExtensionsGetAllDevicesAsync
+{
+    class Program
+    {
+        static async Task Main(string[] args)
+        {
+            using (var adapter = (await BlueZManager.GetAdaptersAsync()).FirstOrDefault())
+            {
+                try
+                {
+                    if (adapter != null)
+                    {
+                        var props = new Dictionary<string, object>();
+                        props.Add("Transport", "le");
+                        props.Add("DuplicateData", false);
+                        await adapter.SetDiscoveryFilterAsync(props);
+                        await adapter.StartDiscoveryAsync();
+                        Console.WriteLine("Waiting for devices. Use Control-C to quit.");
+                        while (true)
+                        {
+                            var devices = await adapter.GetDevicesAsync();
+                            foreach (var device in devices)
+                            {
+                                var deviceProperties = await device.GetAllAsync();
+                                Console.WriteLine($"{deviceProperties.Name} ({deviceProperties.Address}) was the device.");
+                            }
+                        }
+                    }
+                    else
+                    {
+                        Console.Error.WriteLine("Failed to get the Adapter; Exiting");
+                        Environment.ExitCode = -1;
+                    }
+                }
+                catch (Exception ex)
+                {
+                    Console.Error.WriteLine(ex);
+                }
+                finally
+                {
+                    Console.WriteLine("Stopping Discovery");
+                    if (adapter != null)
+                    {
+                        await adapter.StopDiscoveryAsync();
+                    }
+                }
+            }
```

Comment:
Okay i have some suggestions for this code.

You can use a `using` statement without curly brackets, like this `using var adapter = (await BlueZManager.GetAdaptersAsync()).FirstOrDefault();` also called a simple using statement. This will limit the nesting and make it easier to read. It makes sense to use a simple using statement here as we are using the object throughout the method

You can also reverse the `if (adapter != null)` to `if (adapter == null)` and run the "else" code in that if block. This will eliminate a level of nesting again.

If we apply my suggestions the code will look like this:

```csharp
using var adapter = (await BlueZManager.GetAdaptersAsync()).FirstOrDefault();
if (adapter != null)
{
    Console.Error.WriteLine("Failed to get the Adapter; Exiting");
    Environment.ExitCode = -1;
}
try
{
    var props = new Dictionary<string, object>();
    props.Add("Transport", "le");
    props.Add("DuplicateData", false);
    await adapter.SetDiscoveryFilterAsync(props);
    await adapter.StartDiscoveryAsync();
    Console.WriteLine("Waiting for devices. Use Control-C to quit.");
    while (true)
    {
        var devices = await adapter.GetDevicesAsync();
        foreach (var device in devices)
        {
            var deviceProperties = await device.GetAllAsync();
            Console.WriteLine($"{deviceProperties.Name} ({deviceProperties.Address}) was the device.");
        }
    }
}
catch (Exception ex)
{
    Console.Error.WriteLine(ex);
}
finally
{
    Console.WriteLine("Stopping Discovery");
    if (adapter != null)
    {
        await adapter.StopDiscoveryAsync();
    }
}
```

Reply:
Huh TIL; specifically https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/style-rules/ide0063 which was introduced in C# 8.

The changes you suggest above I have applied; I believe that it required other minor changes to the control flow (the way this was written before it only had a single return point). Now if it errors in the `if(adapter == null)` check I simply `return` after setting the `Environment.ExitCode`.

I also believe that your suggestions also avoid the need for the `if(adapter != null)` check in the `finally` statement and have updated the program to reflect that. If this was wrong please let me know! I appreciate your time to perform a review.

Resolved on [cac7288](https://github.com/vestervang/DotNet-BlueZ/pull/10/commits/cac72886517e1d7aaf799dcdc3fd2d3f3135527c)